### PR TITLE
Add --openshift-sar-by-host option to specify SAR rules per hostname

### PR DESCRIPTION
**Pull request by kincl**
This builds on --openshift-sar by having rules that apply to a specific hostname. In this way oauth-proxy can be used for multiple upstream hosts with different authorization rules. 

In doing this work I realized that the cookie once issued would work for any URL so I append the hostname to the cookie secret so that the cookie only validates for the hostname that it was issued with.

### CODE REVIEW

**Comment by simo5** [comment]
@mrogers950 what do you think ?
My fear is that we'll have to add -by-host to more commands down the road and we end up with complications, but I am not opposed in principle.

File: README.md [hunk 1]
```diff
@@ -69,13 +69,24 @@ Example:
...
 
 Run `oc explain subjectaccessreview` to see the schema for a review, including other fields.
 Specifying multiple rules via a JSON array (`[{...}, {...}]`) will require all permissions to
 be granted.
 
+##### Require specific permissions per host with `--openshift-sar-by-host=JSON`
+
+This is similar to the `--openshift-sar` option but instead of the rules applying to all hosts, you
+can set up specific rules that are checked for a particular upstream host. Using a JSON object the
+keys are hostnames and the value is a JSON array of SAR rules.
```
Comment (mrogers950) [inline README.md]: How does this work alongside --openshift-sar? Will --openshift-sar= compound on each host-specific sar?

Comment (mrogers950) [reply README.md]: If they work together in that way, it might then be better to just change --openshift-sar to allow in its array a combination of global rules and host-specific rules.

Comment (kincl) [reply README.md]: Yes they are just compounded together. I was initially going down that path but I couldn't figure out the best way to distinguish between a single SAR rule object and an object mapping of hosts to SAR rule or array of SAR rules.

File: README.md [hunk 2]
```diff
@@ -107,18 +118,18 @@ Example:
...
 may be unwittingly provide their credentials to untrusted components that can then act as them.
 
 
 ### Other configuration flags
 
 #### `--openshift-service-account=NAME`
 
-Will attempt to read the `--client-id` and `--client-secret` from the service account information 
+Will attempt to read the `--client-id` and `--client-secret` from the service account information
 injected by OpenShift. Uses the value of `/var/run/secrets/kubernetes.io/serviceaccount/namespace`
```
Comment (simo5) [inline README.md]: nit: can you avoid touching unrelated code/documentation lines ?

Comment (kincl) [reply README.md]: Sure thing, just my editor removing extra spaces. Done.

File: README.md (hunk 1 above)
Comment (mrogers950) [reply README.md]: OK, please add a mention that both options work together in the README.

**Comment by enj** [comment]
Isn't this starting to get a bit complicated?

**Comment by kincl** [comment]
@enj yes you could be correct if oauth-proxy never plans to support multiple hosts in upstream. If we are not precluding multiple hosts in upstream, we are already doing authentication with oauth-proxy so it makes it a perfect place to do per-host authorization as well.

Are you concerned about the complexity of the similar-but-different option? I could go back over and try doing this with just `--openshift-sar`.

File: README.md (hunk 1 above)
Comment (kincl) [reply README.md]: Done

**Comment by mrogers950** [comment]
@kincl please squash the commits - I'm OK with the option in its current form. @simo5 or @enj do you have any further feedback?

**Comment by simo5** [comment]
SGTM
